## Problem #158: Lexicographical Neighbours
[Link to Problem](https://projecteuler.net/problem=158)

### Problem Description

Taking three different letters from the $26$ letters of the alphabet, character strings of length three can be formed.

Examples are 'abc', 'hat' and 'zyx'.

When we study these three examples we see that for 'abc' two characters come lexicographically after its neighbour to the left.

For 'hat' there is exactly one character that comes lexicographically after its neighbour to the left. For 'zyx' there are zero characters that come lexicographically after its neighbour to the left.
In all there are $10400$ strings of length $3$ for which exactly one character comes lexicographically after its neighbour to the left.

We now consider strings of $n \leq 26$ different characters from the alphabet.
For every $n$, $p(n)$ is the number of strings of length $n$ for which exactly one character comes lexicographically after its neighbour to the left.

What is the maximum value of $n$?

### Approach

Let $\sigma = 26$ (the size of the alphabet).

First, we can turn every string into a permutation according to the relative order of the letters. Thus $abc$ becomes $123$, $hat$ becomes $213$ and $zyx$ becomes $321$.

Then if we can compute $f(n)$ = the number of valid permutations, $p(n)$ simply becomes $f(n) \times \binom{\sigma}{n}$.

We can further encode each permutation of length $n$ by a string of length $n - 1$ of `'<'` and `'>'`: $123$ turns into `"<<"`, $213$ turns into `"><"`, etc.

Now $f(n)$ becomes the number of permutations whose encoding contains **exactly** one occurence of `'<'`. To do this we generate each possible encoding and for a fixed encoding we can find the number of permutations that fit as explained in [this blog](https://codeforces.com/blog/entry/125175).

In [21]:
# Combinatorics prerequisites
sigma = 26

factorial = [1] * (sigma + 1)

for i in range(2, sigma + 1):
    factorial[i] = factorial[i - 1] * i

def binom(n, k):
    return factorial[n] // factorial[k] // factorial[n - k]

In [31]:
# Count the number of possible permutations fit the encoding
def compute(encoding: str):
    length = len(encoding) + 1
    dp = [[0 for j in range(i + 1)] for i in range(length)]
    # dp[i][j] = how many 0-indexed permutations of length i + 1
    #            have the last element equal to j (j <= i)
    dp[0][0] = 1
    for i in range(1, length):
        for j in range(i + 1):
            if encoding[i - 1] == '>':
                for k in range(j, i):
                    dp[i][j] += dp[i -  1][k]
            else:
                for k in range(j):
                    dp[i][j] += dp[i -  1][k]
    return sum(dp[-1])

In [ ]:
limit = 26

f = [0] * (limit + 1)

for length in range(2, limit + 1):
    encoding = '>' * (length - 1)
    for i in range(length - 1):
        new_encoding = encoding[:i] + '<' + encoding[i + 1:]
        ways = compute(new_encoding)
        # print(new_encoding, ways)
        f[length] += ways

p = [value * binom(sigma, index) for index, value in enumerate(f)]

print(max(p))

409511334375


###### Result: **409511334375** | Execution time: 0s

### Complexity analysis

- $\sigma$ values for the length
- $\sigma$ encodings for a fixed length
- $\sigma^2$ algorithm to count the permutations for an encoding

$\Rightarrow$ Time complexity: $O(\sigma^4)$

##### Tags: #strings, #combinatorics, #dynamic-programming